src/python/pants/pantsd/pailgun_server.py

```diff
@@ -90,13 +90,15 @@ def handle_error(self, exc=None):
 class PailgunServer(TCPServer):
   """A (forking) pants nailgun server."""
 
-  def __init__(self, server_address, runner_factory, handler_class=None, bind_and_activate=True):
+  def __init__(self, server_address, runner_factory, lifecycle_lock,
+               handler_class=None, bind_and_activate=True):
     """Override of TCPServer.__init__().
 
     N.B. the majority of this function is copied verbatim from TCPServer.__init__().
 
     :param tuple server_address: An address tuple of (hostname, port) for socket.bind().
     :param class runner_factory: A factory function for creating a DaemonPantsRunner for each run.
+    :param threading.RLock lifecycle_lock: A lifecycle lock to prevent abrupt teardown during handling.
```

Comment:
Naming is hard, but: it's not clear _which_ lifecycle is covered by this lock. Might be good to mention services, or refer to where the lock is created and put a larger docstring there.